File: packages/api/src/scripts/migrate.ts
#!/usr/bin/env node
/**
 * Database migration CLI tool
 * 
 * Usage:
 *   npm run migrate status          - Show migration status
 *   npm run migrate up              - Apply all pending migrations
 *   npm run migrate down            - Rollback last migration
 *   npm run migrate to <version>    - Migrate to specific version
 */

import { connectMongoDB, closeMongoDB } from '../database/mongodb';
import { MigrationManager, allMigrations } from '../migrations';

async function main() {
  const command = process.argv[2];
  const arg = process.argv[3];

  try {
    const db = await connectMongoDB();
    const manager = new MigrationManager(db, allMigrations);

    switch (command) {
      case 'status': {
        const status = await manager.getStatus();
        console.log('\n=== Migration Status ===');
        console.log(`Current version: ${status.currentVersion}`);
        console.log(`Latest version: ${status.latestVersion}`);
        console.log(`\nApplied migrations (${status.appliedMigrations.length}):`);
        for (const migration of status.appliedMigrations) {
          console.log(
            `  ${migration.version}: ${migration.name} (${migration.appliedAt.toISOString()})`
          );
        }
        console.log(`\nPending migrations (${status.pendingMigrations.length}):`);
        for (const migration of status.pendingMigrations) {
          console.log(`  ${migration.version}: ${migration.name}`);
        }
        break;
      }

      case 'up': {
        await manager.migrateToLatest();
        break;
      }

      case 'down': {
        await manager.rollbackLast();
        break;
      }

      case 'to': {
        if (!arg) {
          console.error('Error: Version number required');
          console.log('Usage: npm run migrate to <version>');
          process.exit(1);
        }
        const version = parseInt(arg, 10);
        if (isNaN(version)) {
          console.error('Error: Invalid version number');
          process.exit(1);
        }
        await manager.migrateTo(version);
        break;
      }

      default: {
        console.log('Database Migration Tool\n');
        console.log('Usage:');
        console.log('  npm run migrate status          - Show migration status');
        console.log('  npm run migrate up              - Apply all pending migrations');
        console.log('  npm run migrate down            - Rollback last migration');
        console.log('  npm run migrate to <version>    - Migrate to specific version');
        break;
      }
    }

    await closeMongoDB();
    process.exit(0);
  } catch (error) {
    console.error('Migration failed:', error);
    await closeMongoDB();
    process.exit(1);
  }
}

main();

File: packages/agents/src/scripts/migrate.py
#!/usr/bin/env python3
"""
Database migration CLI tool

Usage:
    python -m src.scripts.migrate status          - Show migration status
    python -m src.scripts.migrate up              - Apply all pending migrations
    python -m src.scripts.migrate down            - Rollback last migration
    python -m src.scripts.migrate to <version>    - Migrate to specific version
"""
import sys
import asyncio

from ..database.mongodb import connect_mongodb, close_mongodb
from ..migrations import MigrationManager, all_migrations


async def main():
    """Main entry point"""
    if len(sys.argv) < 2:
        print_usage()
        return

    command = sys.argv[1]

    try:
        db = await connect_mongodb()
        manager = MigrationManager(db, all_migrations)

        if command == "status":
            await show_status(manager)
        elif command == "up":
            await manager.migrate_to_latest()
        elif command == "down":
            await manager.rollback_last()
        elif command == "to":
            if len(sys.argv) < 3:
                print("Error: Version number required")
                print("Usage: python -m src.scripts.migrate to <version>")
                sys.exit(1)
            try:
                version = int(sys.argv[2])
                await manager.migrate_to(version)
            except ValueError:
                print("Error: Invalid version number")
                sys.exit(1)
        else:
            print_usage()

        await close_mongodb()
    except Exception as e:
        print(f"Migration failed: {e}")
        await close_mongodb()
        sys.exit(1)


async def show_status(manager: MigrationManager):
    """Show migration status"""
    status = await manager.get_status()

    print("\n=== Migration Status ===")
    print(f"Current version: {status['current_version']}")
    print(f"Latest version: {status['latest_version']}")

    print(f"\nApplied migrations ({len(status['applied_migrations'])}):")
    for migration in status["applied_migrations"]:
        print(
            f"  {migration.version}: {migration.name} "
            f"({migration.applied_at.isoformat()})"
        )

    print(f"\nPending migrations ({len(status['pending_migrations'])}):")
    for migration in status["pending_migrations"]:
        print(f"  {migration.version}: {migration.name}")


def print_usage():
    """Print usage information"""
    print("Database Migration Tool\n")
    print("Usage:")
    print("  python -m src.scripts.migrate status          - Show migration status")
    print("  python -m src.scripts.migrate up              - Apply all pending migrations")
    print("  python -m src.scripts.migrate down            - Rollback last migration")
    print("  python -m src.scripts.migrate to <version>    - Migrate to specific version")


if __name__ == "__main__":
    asyncio.run(main())
